parse_node/parse_node_with_comments.ts

```diff
@@ -0,0 +1,180 @@
+import ts from "typescript"
+import {
+  ParseNodeType,
+  parseNodeWithoutComments,
+  ParseState,
+} from "../parse_node"
+import { Test } from "../tests/test"
+
+const getBackwardsClosestNewLinesCount = (text: string, pos: number) => {
+  let count = 0
+
+  for (let i = pos - 1; i >= 0; i--) {
+    const char = text.charAt(i)
+    if (char === "\t" || char === " ") {
+      continue
+    }
+
+    if (char === "\n") {
+      count++
+      continue
+    }
+
+    break
+  }
+
+  return count
+}
+
+export const parseNodeWithComments = (
+  node: ts.Node,
+  props: ParseState
+): ParseNodeType => {
+  if (ts.isSourceFile(node)) {
+    return parseNodeWithoutComments(node, props)
+  }
+
+  let leadingComments =
+    ts
+      .getLeadingCommentRanges(
+        node.getSourceFile().getFullText(),
+        node.getFullStart()
+      )
+      ?.filter((v) => v.kind === ts.SyntaxKind.SingleLineCommentTrivia) ?? []
+
+  let tarilingComments =
+    ts
+      .getTrailingCommentRanges(
+        node.getSourceFile().getFullText(),
+        node.getFullStart() + node.getFullWidth()
+      )
+      ?.filter((v) => v.kind === ts.SyntaxKind.SingleLineCommentTrivia) ?? []
+
+  let commentStackUnwind = props.commentsStack?.length ?? 0
+
+  if (leadingComments.length > 0 || tarilingComments.length > 0) {
+    if (!props.commentsStack) {
+      props.commentsStack = []
+    }
+
+    leadingComments = leadingComments.filter(
+      (v) =>
+        !props.commentsStack!.find((c) => c.pos === v.pos && c.end === v.end)
```

Comment:
I believe you that we need to dedupe them, but why? Does TS give us back the same comment multiple times? When and why? 

It might help to leave a comment about this.

Reply:
The problem is that multiple nodes can give the same comment when they are in the same line. For example:

```ts
// some comment
myFunc(1 + 2, someVar)
```

This code generates AST like: `CallExpression (params: BinaryExpression (left: NumberLiteral (1), right: NumberLiteral (2)), Identifier (someVar))`. All nodes in this AST will report `// some comment` that is why we need to handle only first node (CallExpression) that report this comments and ignore the rest. I will add some comments.

Reply:
Ah, I understand. Bit of a wild guess here, but maybe comments are mapped 1:1 with statements, and then copied into child nodes? Could you try just checking if node is a statement type and ONLY if it is appending the comments?

e.g. like this:
https://github.com/johnfn/ts2gd/blob/main/parse_node.ts#L169

Reply:
Because it seems if we could solve this then a lot of this code would be simplified - we wouldn't even need to track comments at all in the ParseState!

Reply:
I tried limiting it only to statements but then cases like one below are removed. I don't think comments are mapped 1:1 with statements because when single statement is divided to multiple lines and there are comments inside those lines then comments are bound to expression nodes and the statement.
```ts
someCall(
  // parameter comment
  "abcd",
  // next comment
 123
)
```

```gdscript
someCall("abcd", 123)
```

I also tested your example from other comment and indeed it generates invalid GDScript code, so I think we somehow handle these weird cases or we do as you said: limit comment generation only for statements and this way we won't need too worry about such cases.

```ts
// example from other comment
let foo = { bar: 1 }
console.log(foo. 
  // Hello
bar)
```
I don't really know which way to move forward so I'll wait for your opinion. Also there is not so good another way, just leave it as it is and if someone creates comment that generates invalid GDScript then Godot will actually tell him that something is wrong anyway.